mindmeld/resource_loader.py

```diff
@@ -40,7 +40,176 @@
+        '''
+        return ProcessedQueryList(
+            cache=ProcessedQueryList.MemoryCache(queries),
+            row_ids=list(range(len(queries)))
+        )
+
+    class Iterator:
+        def __init__(self, source, cached=False):
+            self.source = source
+            self.row_ids = source.row_ids
+            self.result_cache = [] if not cached else [None] * len(source)
+            self.iter_idx = -1
+
+        def __iter__(self):
+            self.iter_idx = -1
+            return self
+
+        def __next__(self):
+            self.iter_idx += 1
+            if self.iter_idx == len(self):
+                raise StopIteration
+            if self.result_cache and self.result_cache[self.iter_idx] is not None:
+                return self.result_cache[self.iter_idx]
+            result = self[self.iter_idx]
+            if self.result_cache:
+                self.result_cache[self.iter_idx] = result
+            return result
+
+        def __len__(self):
+            return len(self.source)
+
+        def __getitem__(self, key):
+            '''
+            Override this function for getting other types of cached data.
+            '''
+            return self.source.cache.get(self.row_ids[key])
+
+        def reorder(self, indices):
+            self.row_ids = [self.row_ids[i] for i in indices]
+            if self.result_cache:
+                self.result_cache = [self.result_cache[i] for i in indices]
+
+    class RawQueryIterator(Iterator):
+        def __getitem__(self, key):
+            return self.source.cache.get_raw_query(self.row_ids[key])
+
+    class QueryIterator(Iterator):
+        def __getitem__(self, key):
+            return self.source.cache.get_query(self.row_ids[key])
+
+    class EntitiesIterator(Iterator):
+        def __getitem__(self, key):
+            return self.source.cache.get_entities(self.row_ids[key])
+
+    class DomainIterator(Iterator):
+        def __init__(self, source):
+            cached = not isinstance(source.cache, ProcessedQueryList.MemoryCache)
+            super().__init__(source, cached=cached)
+
+        def __getitem__(self, key):
+            return self.source.cache.get_domain(self.row_ids[key])
+
+    class IntentIterator(Iterator):
+        def __init__(self, source):
+            cached = not isinstance(source.cache, ProcessedQueryList.MemoryCache)
+            super().__init__(source, cached=cached)
+
+        def __getitem__(self, key):
+            return self.source.cache.get_intent(self.row_ids[key])
+
+    class ListIterator(Iterator):
+        def __init__(self, elements):
+            self.source = None
+            self.row_ids = None
+            self.result_cache = None
+            self.iter_idx = -1
+            self.elements = elements
+
+        def __getitem__(self, key):
+            return self.elements[key]
+
+        def __len__(self):
+            return len(self.elements)
+
+        def reorder(self, indices):
```

Comment:
Cant we just use the parent class's `reorder ` since `result_cache ` is None anyway?